content/6_Advanced/random.mdx

```diff
@@ -0,0 +1,335 @@
+		ch = getchar();
+	}
+	return num * sign;
+}
+
+mt19937 rng(chrono::steady_clock::now().time_since_epoch().count());
+unsigned long long n, a[mxn], dv[mxn], cnt[mxn], ans, x, ind, d;
+
+signed main() {
+	ios::sync_with_stdio(false);
+	cin.tie(nullptr);
+	n = read();
+	for (int i = 0; i < n; i++) a[i] = read();
+	for (int s = 0; s < S; s++) {
+		x = a[rng() % n], ind = 0;
+		for (d = 1; d * d <= x; d++)
+			if (x % d == 0) {
+				dv[ind++] = d;
+				if (d * d != x) dv[ind++] = x / d;
+			}
+		sort(dv, dv + ind);
+		fill(cnt, cnt + ind, 0);
+		for (int i = 0; i < n; i++)
+			cnt[lower_bound(dv, dv + ind, __gcd(a[i], x)) - dv]++;
+		for (int i = 0; i < ind; i++)
+			for (int j = 0; j < i; j++)
+				if (dv[i] % dv[j] == 0) cnt[j] += cnt[i];
+		for (int i = 0; i < ind; i++)
+			if (cnt[i] + cnt[i] >= n) ans = max(ans, dv[i]);
+	}
+	cout << ans;
+}
+```
+
+## Example - Count on a Tree II Striking Back
+
+<FocusProblem problem="coat2sb" />
+
+Trying to solve this problem directly - explicitly finding values for $f(a,b)$ and $f(c,d)$ - is quite difficult. Supporting online queries of range count distinct along with updates is quite difficult to achieve on an array in reasonable complexity, and the fact that this problem occurs on a tree, with relative large bound for $n \leq 500$ $000$ makes it clear that solving for the value of $f$ is not intended.
+
+Instead, we can attempt to use the property given in the statement that either $f(a,b) \geq 2f(c,d)$ or $f(c,d) \geq 2f(a,b)$. This inspires the following thought - if we map each color to a random number, the path minimum along the path with more distinct colors will *probably* be a smaller value.
+
+We attempt to formalize this approach, and push the probability to an acceptable bound. Notably, the probability of failure is too high when we do the aforementioned random process once. To remedy this, we proceed as follows:
+
+- First we fix $S$ to be the number of "simulations" that we run.
+- Next, for each possible color $1 \leq c \leq n$, we define $v_{cj}$ as an independent random value within $[0,1]$, for each $1 \leq j \leq S$.
+- Then, we construct $S$ trees $T_i$, of the same structure as the initial tree, but with $v_{c_u i}$ marked on a node $u$.
+- For each query comparing $a,b,c,d$, we output $\texttt{Yes}$ if
+$$
+\sum_{i\leq S} \text{PathMin}(T_i,a,b) < \sum_{i\leq S} \text{PathMin}(T_i,c,d)
+$$
+and $\texttt{No}$ otherwise.
+
+This approach should "combine" the probabilities in a way that makes the more likely event (the path with more distinct values achieving a smaller value) exponentially more likely. For some intuition on this, if we flip a coin weighted with $\frac{2}{3}$ probability to be heads once, there is a $\frac{1}{3}$ probability it comes up tails more than heads. However, if we flip it $999$ times, the probability it comes up more tails more than heads becomes
+
+$$
+\sum_{k=50}^{99} \binom{99}{k} \left(\frac{1}{3}\right)^k \left(\frac{2}{3}\right)^{99-k} \approx .000309
```

Comment:
this seems like quite a high failure probability. considering that there are 10K queries ...

Reply:
ok I guess summing the minimum is better than summing coin tosses

Reply:
> probability exponentially decreasing with increased trials.

it's not exponential is it? it's more like e^{-sqrt trials}

Reply:
> Should I just remove the coin toss example completely if it is too distracting?

well, it's better than nothing but I feel like the intuition it gives is also sorta misleading

Reply:
from the coin toss example you presented, it gives me the intuition that randomness is not sufficient to solve this problem :/

Reply:
I am also confused by the graph. Why are you only using S=75 if it only lets you pass a test case with better than half probability?

Reply:
>  In fact, you can get AC on this problem even with S=28...

huh the test data must be terrible then